cs/src/core/Index/FASTER/FASTERImpl.cs

```diff
@@ -1439,6 +1509,33 @@ private void BlockAllocate(int recordSize, out long logicalAddress)
             }
         }
 
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void BlockAllocateReadCache(int recordSize, out long logicalAddress)
+        {
+            logicalAddress = readcache.Allocate(recordSize);
+            if (logicalAddress >= 0)
+                return;
+
+            while (logicalAddress < 0 && -logicalAddress >= hlog.ReadOnlyAddress)
+            {
+                InternalRefresh();
+                hlog.CheckForAllocateComplete(ref logicalAddress);
+                if (logicalAddress < 0)
+                {
+                    Thread.Sleep(10);
+                }
+            }
+
+            logicalAddress = logicalAddress < 0 ? -logicalAddress : logicalAddress;
+
+            if (logicalAddress < hlog.ReadOnlyAddress)
+            {
+                Debug.WriteLine("Allocated address is read-only, retrying");
+                BlockAllocateReadCache(recordSize, out logicalAddress);
```

Comment:
>BlockAllocateReadCache(recordSize, out logicalAddress); [](start = 16, length = 55)

Does this need to be recursive? What is preventing this from looping infinitely?

Reply:
This catches the case where ReadOnlyAddress is updated after allocation but before the allocated address becomes available. Because ReadOnlyAddress is incremented in coarse-grained chunks, it will not recurse forever. Also note that this is tail recursion as before.